### ISSUE TITLE
Tuple reducers

### ISSUE THREAD

**Issue by isiahmeadows** [Microsoft/TypeScript#26058]
## Search Terms

Iterate and merge tuples using recursion and various n-ary reduction operators.

## Suggestion

1. I would like to be able to extract and spread in positions other than the last part of a tuple.
1. I would like to be able to do math on integer literals, for tuple iteration.
1. I would like to merge a tuple into a union, intersection, or concatenation of its members.

This builds on the [mapped tuple types proposal](https://github.com/Microsoft/TypeScript/issues/25947), so it depends on that.

## Use Cases

- Converting a tuple to its union, such as with `Promise.race`'s return value.
- Converting a tuple to its intersection, such as with `Object.assign`'s assignee.
- Typing variadic parameters where the last parameter is a function, but the rest are combined somehow, like in `Observable.combineLatest(...observables, project: (...values) => R): Observable<R>` and as is required for [this bug](https://github.com/Microsoft/TypeScript/issues/1360).
- Typing `Array.prototype.concat` correctly for tuples.

## Syntax

The syntax comes in a few new forms:

- `[...A, B]` appends `B` to the tuple `A`. Similarly, `[...infer A, any] extends B ? A : []` drops the last item of the tuple and `[...any[], infer A] extends B ? A : never` extracts the first.
- `M + N` and `M - N` provide type-level addition and subtraction, for easier array indexing.
- `| ...T` performs an n-ary union, `& ...T` an n-ary intersection, and `[... ...T]` an n-ary concatenation.

## Examples

Here's the types for each method I mentioned above.

```ts
type _ArrayLikeToArray<A, I extends number, R extends any[]> = (
    A["length"] extends I ? I extends A["length"] ? true : false : false
) extends true ? R : [A[I], ...ArrayLikeToArray<A, I + 1, R>];

interface Array<T> {
    concat(...others: any[]): [... ...[(
        N extends any[] ? N : N extends ArrayLike<any> ? _ArrayLikeToArray<N, 0, []> : [N]
    ) for N in typeof others]];
}

interface ObjectConstructor {
    assign(target: & ...typeof sources, ...sources: object[]): typeof target;
}

type Await<N> = N extends PromiseLike<infer M> ? Await<M> : N;

interface PromiseConstructor {
    all(...values: any[]): Promise<[Await<T> for T in typeof values]>;
    race(...values: any[]): Promise<| ...[Await<T> for T in typeof values]>;
}

interface ObservableConstructor {
    combineLatest<V extend any[], R>(...args: [
        ...[Observable<T> for T in V],
        (...values: V) => R,
    ]): Observable<R>;
}
```

Also, there could be a few built-in types:

- `type Indices<T extends any[], I extends number = 0> = I extends T["length"] ? [] : [I, ...Indices<T, I + 1>`
- `type Range<S extends number, E extends number> = S extends E ? [] : [S, ...Range<S + 1, E>]`

And finally, merging is just syntax sugar now: `{...A, ...B}` &harr; `(& ...[A, B])`

## Checklist

My suggestion meets these guidelines:
* [x] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. new expression-level syntax)

**Comment by ajafff** [comment]
> And finally, merging is just syntax sugar now: `{...A, ...B}` ↔ `(& ...[A, B])`

I don't think this is correct. It should behave like object spread in JS and override existing properties.
```ts
type T = {... {a: string}, ...{a: number}};
// should result in
type T = {a: number};
// instead of
type T = {a: string} & {a: number};
```

**Comment by isiahmeadows** [comment]
@ajafff Fixed by removing the line. I forgot about that bit.

**Comment by AlCalzone** [comment]
> I would like to be able to extract and spread in positions other than the last part of a tuple.

**Comment by AlCalzone** [comment]
The 1st part:
> I would like to be able to extract and spread in positions other than the last part of a tuple.

Seems to be a duplicate of: #25717 and therefore #1360

**Comment by isiahmeadows** [comment]
@AlCalzone For the first half, probably, but the second half is likely not.

**Comment by bterlson** [comment]
See also #5453 for part 1.

**Comment by weswigham** [comment]
> Converting a tuple to its union, such as with Promise.race's return value.
Doesn't 
```ts
type TupleTypes<T extends unknown[]> = T[number]
```
?
work for that?

**Comment by isiahmeadows** [comment]
@weswigham Good catch - no clue how that escaped me. I updated the proposal comment to drop that (and those covered by #26063), but the issue with concatenation and `Object.assign` still remains.

I changed up the proposed syntax a little, too, to better fit in:

- Merge all: `{... ...T}`
- Concatenate all: `[... ...T]`

Of course, that space in the middle isn't significant, just explains how it's tokenized.

**Comment by weswigham** [comment]
> Converting a tuple to its intersection, such as with Object.assign's assignee.

Ignoring tuples for a second, I want a way to easily convert any union to an intersection. But y'know, I think it is doable, if cumbersome:

```ts
type MergeArguments<T, K extends string = "whatever"> =
    {
        [Key in K]: T extends (first: infer A) => void ? A :
        MergeOnePlus<T, K>
    }[K];

type MergeOnePlus<T, K extends string> =
    {
        [Key in K]: T extends (first: infer A, ...args: infer U) => void ? A & MergeArguments<(...args: U) => void, K> :
        never
    }[K];

type IntoSignature<T extends unknown[]> = (...args: T) => void;

type MergeTupleMembers<T extends unknown[]> = MergeArguments<IntoSignature<T>>;

type SomeTuple = [{ x: 1 }, { y: 2 }, { z: 3 }];

type Merged = MergeTupleMembers<SomeTuple>;

const x: Merged = {
    x: 1,
    y: 2,
    z: 3
};
```

It's a bit ridiculous, but.... possible.

**Comment by isiahmeadows** [comment]
@weswigham That might very well solve the issue of both concatenation and merging, if I'm reading it correctly.

By any chance, would that still work if you did this to do concatenation?

```ts
type ConcatenateOnePlus<T, K extends string> =
    {
        [Key in K]: T extends (first: infer A, ...args: infer U) => void ? [A, ...ConcatenateArguments<(...args: U) => void, K>] :
        never
    }[K];
```

Or would that be dependent on #25717 / #1360?

**Comment by karol-majewski** [comment]
> Ignoring tuples for a second, I want a way to easily convert any union to an intersection

@weswigham check this out:

```ts
/**
 * @author https://stackoverflow.com/users/2887218/jcalz
 * @see https://stackoverflow.com/a/50375286/10325032
 */
type UnionToIntersection<Union> =
  (Union extends any
    ? (argument: Union) => void
    : never
  ) extends (argument: infer Intersection) => void
      ? Intersection
      : never;
```

**Comment by lieene** [comment]
I'm working on a utility lib with Merg(Intersection), object only Merg, the well-known symbol iterator walkaround. and type morphing. Please check if they are okay. (this is my first ts lib)
``` typescript
export type Omit2<T, K extends keyof T> = T extends { [Symbol.iterator]: infer U } ? { [Symbol.iterator]: U } & Omit<T, K> : Omit<T, K>;
export type ExcludeO<T extends object, U extends object> =
  U extends { [Symbol.iterator]: any } ? Omit<T, keyof U> :
  T extends { [Symbol.iterator]: infer IT } ? { [Symbol.iterator]: IT } & Omit<T, keyof U> : Omit<T, keyof U>;

export type ExtractO<T extends object, U extends object> =
  U extends { [Symbol.iterator]: any } ? T extends { [Symbol.iterator]: infer IT } ?
  { [Symbol.iterator]: IT } & Pick<T, Extract<keyof T, keyof U>> :
  Pick<T, Extract<keyof T, keyof U>> : Pick<T, Extract<keyof T, keyof U>>;

export type Merg<U> = (U extends any ? (k: U) => void : never) extends (k: infer I) => void ? I : never;
export type MergO<U extends object> = (U extends object ? (k: U) => void : never) extends (k: infer T) => void ?
  (T extends object ? T : object) : object;
  
export type UnionTupleType<A extends any[]> = A extends { [n: number]: infer T } ? T extends object ? T : never : never;
export type MergTupleType<A extends any[]> = MergO<UnionTupleType<A>>;

export type Alter<T extends object, U extends object> = ExcludeO<T, U> & ExtractO<U, T>;
export type Extend<T extends object, U extends object> = T & ExcludeO<U, T>;
export type Override<T extends object, U extends object> = ExcludeO<T, U> & U;

export type AlterOver<T extends object, U extends object, X extends object> = Alter<T, ExcludeO<U, X>>;
export type ExtendOver<T extends object, U extends object, X extends object> = Extend<T, ExcludeO<U, X>>;
export type OverrideOver<T extends object, U extends object, X extends object> = Override<T, ExcludeO<U, X>>;

export type AlterLike<T extends object, U extends object, X extends object> = Alter<T, ExtractO<U, X>>;
export type ExtendLike<T extends object, U extends object, X extends object> = Extend<T, ExtractO<U, X>>;
export type OverrideLike<T extends object, U extends object, X extends object> = Override<T, ExtractO<U, X>>;
```

**Comment by lieene** [comment]
@weswigham this should fit your need and it enforces object
and they are [here](https://www.npmjs.com/package/@lieene/ts-utility)
```typescript
export type MergO<U extends object> = (U extends object ? (k: U) => void : never) extends (k: infer T) => void ? (T extends object ? T : object) : object;
export type UnionTupleType<A extends any[]> = A extends { [n: number]: infer T } ? T extends object ? T : never : never;
export type MergTupleType<A extends any[]> = MergO<UnionTupleType<A>>;
```

**Comment by ericbiewener** [comment]
@lieene That is wild! What is going on with the functions in `MergO`? It sort of seems like it's being used to funnel the `U` into `infer T`, I guess because you can't use `infer` inside the branches of a conditional?
